### PR TITLE
replace CASE expressions in predicate pruning with boolean algebra

### PR DESCRIPTION
**Pull request by adriangb**
Boolean algebra is simpler to evaluate and if the metadata is being stored in something that supports indexes e.g. Postgres then this new version can be pushed down to the index while the `CASE` expressions cannot.

This replaces the produced predicate pruning expressions which will change from something along the lines of:

```sql
case when row_count = null_count then false else col_min > 1 end
```

To:

```sql
row_count != null_count and col_min > 1
```

### CODE REVIEW

**Comment by adriangb** [comment]
cc @appletreeisyellow @alamb to understand if there was a reason to make these CASE statements instead of boolean algebra that I am not picking up on

**Comment by adriangb** [comment]
Things that would need to be done before merging this:
- Replace a lot of comments referencing `CASE`
- Reformat the test assertions that are now very long lines (manual work 😢 )

**Comment by findepi** [comment]
@adriangb can you please expand PR description explaining or exemplifying what's replaced with what?

**Review by alamb — COMMENTED** [review]
Thanks @adriangb  -- this change makes lots of sense to me

I think the reason @appletreeisyellow and I used `CASE` before was to avoid introducing a null into the PruningPredicate . 

According to 
https://docs.rs/datafusion/latest/datafusion/physical_optimizer/pruning/struct.PruningPredicate.html#pruningpredicate-implementation

The Pruning predicate should always evaluate to non-null
> Outputs: A (non null) boolean value for each container:
> - true: There MAY be rows that match the predicate
> - false: There are no rows that could possibly match the predicate (the predicate can never possibly be true). The container can be pruned (skipped) entirely.


To compare the `CASE` construct with `NOT` you get

| `null_count = row_count` | `CASE ...` | `NOT null_count = row_count` |
|--------|--------|--------|
| `true` | `false` | `false` | 
| `false` | <rest of expr> | `<rest of expr> |
| `null` | <rest of expr> | <null> | 

If you use `CASE .. ` and there is no row count information, `null_count = row_count` is `null` and the expression simplifies to whatever the rest of the expression evaluates to.

I think if you use `NOT` and there is no row count information, the expression simplifies to `null AND <rest>` 

But `null AND <rest>`


```sql
postgres=# select null AND true;
 ?column?
----------

(1 row)

postgres=# select null AND false;
 ?column?
----------
 f
(1 row)

postgres=# select null AND null;
 ?column?
----------

(1 row)
```


# Suggestions
1. It is concerning that there is no test case that fails in this case (so maybe my analysis is wrong or something else is missing). We should try to add some coverage
2. Can you use `coalesce` instead of `case` to substitute in the `true` (might be matching rows)? For example

```sql
COALESCE(NOT, null_count = row_count, true)
```

**Comment by alamb** [comment]
> I am concerned that the purported benefit
> 
> > something that supports indexes e.g. Postgres then this new version can be pushed down to the index while the CASE expressions cannot.
> 
> may be a result of overly aggressive pruning. I would assume PostgreSQL can handle and optimize CASE expressions the same way or better than we can (but i may be wrong), so CASE vs coalesce shouldn't make a difference.

For some backstory (that I happen to know) is that I believe @adriangb 's usecase is to use the PruningPredicate rewriter to rewrite expressions into range analysis expressions that are then pushed into postgres for execution (aka he is effectively executing the rewritten predicate expression on a posgres database that stores the min/maxes)

I agree it would be useful to understand why case is causing a problem

**Comment by adriangb** [comment]
> 2\. Can you use `coalesce` instead of `case` to substitute in the `true` (might be matching rows)?

That should work:

```sql
postgres=# create index on stats (coalesce(not row_count = null_count, true), min);
CREATE INDEX
postgres=# explain select * from stats where coalesce(not row_count = null_count, false) and min > 599;
                                       QUERY PLAN
-----------------------------------------------------------------------------------------
 Index Scan using stats_min_idx on stats  (cost=0.57..3637302.07 rows=49999713 width=16)
   Index Cond: (min > 599)
   Filter: COALESCE((row_count <> null_count), false)
```

**Comment by alamb** [comment]
I thought more about this in the 🚿  and doubled checked the logic:

TLDR is I think using `NOT` (as in this PR) is ok -- the rationale is that if evaluating `column_count = null_count` is null it means nothing is known about the null_counts. However, since `null AND ...` will still resolve to `false` if the `...` is false (aka we can prove the predicate is not true by other means), then the requirements of the pruning predicate will be satisfied

https://github.com/apache/datafusion/blob/e665115893e6282d592df71657e9f5b5855d1617/datafusion/physical-optimizer/src/pruning.rs#L723-L726

So TLDR is upon more thought I think the theory behind this PR is sound ✅ 

I need to review the code more carefully and ensure we have a test that has unknown (unspecified) column count but the value can be proven true by other min/max ranges but otherwise it should be good to go.

File: datafusion/physical-optimizer/src/pruning.rs
Comment (adriangb) [inline datafusion/physical-optimizer/src/pruning.rs:2113]: @alamb see test case here for https://github.com/apache/datafusion/pull/13795#issuecomment-2548414526

File: datafusion/physical-optimizer/src/pruning.rs
Comment (adriangb) [inline datafusion/physical-optimizer/src/pruning.rs:2118]:
```suggestion
        // If the min or max stats are missing we should not prune
        // (unless we know all rows are null, see `prune_all_rows_null_counts`)
```

**Comment by adriangb** [comment]
Can we go one step further and instead of `not row_count = null_count` use `row_count != null_count`?

**Comment by findepi** [comment]
> `not row_count = null_count` use `row_count != null_count`?

AFAICT those two are equivalent unless operands are IEEE 754 floating point numbers. are they?

File: datafusion/physical-optimizer/src/pruning.rs
Comment (adriangb) [inline datafusion/physical-optimizer/src/pruning.rs:293]: Because of https://github.com/apache/datafusion/blob/f4e65d2d9711ed097982d2fbde4191c402c05023/datafusion/physical-optimizer/src/pruning.rs#L727-L730

**Comment by adriangb** [comment]
I wanted to check that the assumption that falling back on min/max stats when row / null counts are missing/unknown is safe _even when all rows are null_ and _we don't know that all of the rows are null_.

I was afraid that writers might populate the statistics with a default value (e.g. `0`) if all of the rows are null instead of `null`.
And that some other pass might then remove the information on row count / null counts resulting in:

| row_count | null_count | min | max |
|-----------|------------|-----|-----|
| null      | null       | 0   | 0   |

I believe both the previous implementation and this implementation would decide to exclude the container for the predicate `col > 1`:

```sql
CASE WHEN null = null THEN false ELSE col_min > 1 END`
```
```sql
null != null AND col_min > 1
```

Both of these cases return `false` indicating that *we know we can prune the container*.
But that would be wrong!

I tested using our writer:

```rust
use std::sync::Arc;

use parquet::arrow::ArrowWriter;

#[tokio::main]
async fn main() {
    let array = arrow::array::Int32Array::from(vec![None]);
    let array = Arc::new(array);
    let schema = arrow::datatypes::Schema::new(vec![arrow::datatypes::Field::new("a", arrow::datatypes::DataType::Int32, true)]);
    let schema = Arc::new(schema);
    let batch = arrow::record_batch::RecordBatch::try_new(schema.clone(), vec![array]).unwrap();

    let mut file = std::fs::File::create("test.parquet").unwrap();
    let mut writer = ArrowWriter::try_new(&mut file, schema.clone(), None).unwrap();
    writer.write(&batch).unwrap();
    writer.close().unwrap();
}
```

```
❯ parquet meta test.parquet 
Row group 0:  count: 1  38.00 B records  start: 4  total(compressed): 38 B total(uncompressed):38 B 
--------------------------------------------------------------------------------
   type      encodings count     avg size   nulls   min / max
a  INT32     _ RB_     1         38.00 B    1 
```

It does what I think is correct and sets min / max to `null`.
If any writer were to set the min / max values to defaults instead of nulls I think that would be a bug in the writer.
And there would have to be a further bug in the writer or other part of the system to discard or never record the fact that `null_count = row_count`.

So I think this is not a realistic scenario to hit, even if it theoretically could cause issues.

**Comment by findepi** [comment]
> I was afraid that writers might populate the statistics with a default value (e.g. `0`) if all of the rows are null instead of `null`. And that some other pass might then remove the information on row count / null counts resulting in:
> 
> row_count	null_count	min	max
> null	null	0	0

if all values are null, then all non-null values are in [0, 0] range, so min=0 and max=0 would be valid.

**Comment by adriangb** [comment]
So is that assumption incorrect? Then it's a bug in the current implementation.

**Comment by adriangb** [comment]
Thinking about this a bit more it's a real shame these predicates don't return `true` to mean "keep it": if they did the semantics would be compatible with a SQL `where` clause instead of slightly off:
https://github.com/apache/datafusion/blob/f92442ea8e8944c78f8e40d6648d049ff8e335ec/datafusion/physical-optimizer/src/pruning.rs#L261-L263

I suppose it's too late to change that...

**Comment by adriangb** [comment]
> So is that assumption incorrect? Then it's a bug in the current implementation.

Thinking about this again you're right. The worst that can happen is that we return `true` and include the file. If we return `false` well all of the rows are null so we didn't need to scan the container anyway 👍🏻.

**Comment by adriangb** [comment]
In case there is any appetite for it, I found that you can make the predicates always evaluate to `true` or `false` (never `null`) by tacking on a `OR col_a_min IS NULL OR col_a_max IS NULL OR col_a_row_count IS NULL OR col_a_null_count IS NULL` and so on. I think this would be a nice further improvement to make it easier to reason about the logic the predicates return and would be a safee change to make for anyone depending on this code. We already track all referenced columns in `RequiredColumns` 😄

**Comment by adriangb** [comment]
> The only change is for the case where row_count = null_count is null and the predicate is true: previously true would be returned but now null is returned. Since we interpret null as true the end result is the same.

If we wanted to eliminate this change from this PR we can add a condition `and row_count is not null and null_count is not null`. If we go one small step further and add `and col_a_min is not null and col_b_min is not null` then we actually ensure that the predicates always return `true` or `false` and never `null` which IMO would be helpful just to minimize confusion especially since the interpretation of `null` here is so confusing and opposes the interpretation of `null` in a `where` clause.

**Comment by adriangb** [comment]
> If we go one small step further and add `and col_a_min is not null and col_b_min is not null` then we actually ensure that the predicates always return `true` or `false` and never `null` which IMO would be helpful just to minimize confusion especially since the interpretation of `null` here is so confusing and opposes the interpretation of `null` in a `where` clause.

Turns out not that small of a step: there's cases that previously pruned the row that no longer would, in particular when one statistic (`min` _or_ `max`) is missing but the other one isn't and we only use the non-missing statistic.

**Comment by alamb** [comment]
Thank you also @findepi  for your reviews

I'll plan to merge this tomorrow to allow time for other comments unless we find issues or others would like time to review

File: datafusion/physical-optimizer/src/pruning.rs
Comment (appletreeisyellow) [inline datafusion/physical-optimizer/src/pruning.rs:1621]:
```suggestion
/// x_null_count != x_row_count AND (x_min <= 10 AND 10 <= x_max)
```

File: datafusion/physical-optimizer/src/pruning.rs
Comment (appletreeisyellow) [inline datafusion/physical-optimizer/src/pruning.rs:1631]:
```suggestion
    // x_null_count != x_row_count
```

**Review by appletreeisyellow — APPROVED** [review]
Thank you @adriangb for the improvement! 🚀

I also double-checked the logic, and it works as expected. The rewritten pruning predicate is now shorter and easier to read with your change 👏 I also tested this PR against InfluxData IOx code and all the tests pass!

**Comment by alamb** [comment]
I plan to merge this tomorrow morning unless i hear otherwise

**Comment by alamb** [comment]
Thanks again @adriangb and @appletreeisyellow

**Pull request merged**
